trio/_sync.py

```diff
@@ -219,7 +236,7 @@ def acquire_nowait(self):
         self.acquire_on_behalf_of_nowait(trio.lowlevel.current_task())
 
     @enable_ki_protection
-    def acquire_on_behalf_of_nowait(self, borrower):
+    def acquire_on_behalf_of_nowait(self, borrower: Union[object, Task]) -> None:
```

Comment:
`Union[object, Task]` simplifies to `object`

Reply:
@altendky you probably want CapacityLimiter to be generic so you can do:

```suggestion
    def acquire_on_behalf_of_nowait(self, borrower: T) -> None:
```

so you can do:

```
class MyResource:
    def __init__(self):
        self._limit: trio.CapacityLimiter[MyResource] = trio.CapacityLimiter(40)

    def acquire(self):
        self._limit.acquire_on_behalf_of_nowait(self)
    def release(self):
        self._limit.release_on_behalf_of(self)
```

Reply:
the methods that use a task would have to refine the self type:

```python

    @enable_ki_protection
    def acquire_nowait(self: _CapacityLimiter[abc.Task[object]]) -> None:
        """Borrow a token from the sack, without blocking.
        Raises:
          WouldBlock: if no tokens are available.
          RuntimeError: if the current task already holds one of this sack's
              tokens.
        """
        self.acquire_on_behalf_of_nowait(trio.lowlevel.current_task())
```

Reply:
Hints like `Union[object, Task]` were pretty intentional, but open to change.  In this case it mirrored the documentation describing a `Task` as a common particular type of object used here.  In other cases, code actually treated the specific type differently.  Given the suggestion of a more complicated change that I'm not going to get to right now, I'll leave it as is for now.

With the scale of this PR I expect tasks like this to get lost.  I'll make an explicit list of them in the OP with links.  Maybe I'll get lucky and find it is pointlessly redundant.

If there's any interest in bulk contributing to the PR...  I'd welcome it.  `:]`  There's lots more hinting to add, even as just a 'first pass' to satisfy not-so-loose Mypy settings.